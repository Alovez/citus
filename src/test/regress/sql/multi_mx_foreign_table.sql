--
-- multi_mx_foreign_table
-- test foreign tables can be created and queried in mx mode
--

-- create fake fdw for use in tests
CREATE FUNCTION fake_fdw_handler()
RETURNS fdw_handler
AS 'citus'
LANGUAGE C STRICT;

CREATE FOREIGN DATA WRAPPER fake_fdw HANDLER fake_fdw_handler;
CREATE SERVER fake_fdw_server FOREIGN DATA WRAPPER fake_fdw;

-- create fdw on all workers
SELECT FROM run_command_on_workers($$CREATE FUNCTION fake_fdw_handler()
RETURNS fdw_handler
AS 'citus'
LANGUAGE C STRICT; $$);

SELECT FROM run_command_on_workers($$CREATE FOREIGN DATA WRAPPER fake_fdw HANDLER fake_fdw_handler$$);

-- create foreign table
SET citus.shard_replication_factor TO 1;
SET citus.replication_model TO 'streaming';
CREATE FOREIGN TABLE fake_table(a int, b int) SERVER fake_fdw_server;
SELECT create_distributed_table('fake_table', 'a');

-- verify it can be queried from mx node
SELECT * FROM fake_table;
SELECT * FROM run_command_on_workers('SELECT count(*) FROM fake_table')
ORDER BY nodeport;

-- create secont fake table
CREATE FOREIGN TABLE second_fake_table(a int, b int) SERVER fake_fdw_server;
SELECT create_distributed_table('second_fake_table', 'a');

-- verify it can be queried from mx nodes
SELECT * FROM second_fake_table;
SELECT * FROM run_command_on_workers('SELECT count(*) FROM second_fake_table')
ORDER BY nodeport;

-- cleanup resources
DROP FUNCTION IF EXISTS fake_fdw_handler() CASCADE;
SELECT * FROM run_command_on_workers('DROP FUNCTION IF EXISTS fake_fdw_handler() CASCADE')
ORDER BY nodeport;
