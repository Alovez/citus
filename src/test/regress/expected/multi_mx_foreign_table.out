--
-- multi_mx_foreign_table
-- test foreign tables can be created and queried in mx mode
--
-- create fake fdw for use in tests
CREATE FUNCTION fake_fdw_handler()
RETURNS fdw_handler
AS 'citus'
LANGUAGE C STRICT;
CREATE FOREIGN DATA WRAPPER fake_fdw HANDLER fake_fdw_handler;
CREATE SERVER fake_fdw_server FOREIGN DATA WRAPPER fake_fdw;
-- create fdw on all workers
SELECT FROM run_command_on_workers($$CREATE FUNCTION fake_fdw_handler()
RETURNS fdw_handler
AS 'citus'
LANGUAGE C STRICT; $$);
--
(2 rows)

SELECT FROM run_command_on_workers($$CREATE FOREIGN DATA WRAPPER fake_fdw HANDLER fake_fdw_handler$$);
--
(2 rows)

-- create foreign table
SET citus.shard_replication_factor TO 1;
SET citus.replication_model TO 'streaming';
CREATE FOREIGN TABLE fake_table(a int, b int) SERVER fake_fdw_server;
SELECT create_distributed_table('fake_table', 'a');
NOTICE:  foreign-data wrapper "fake_fdw" does not have an extension defined
NOTICE:  foreign-data wrapper "fake_fdw" does not have an extension defined
 create_distributed_table 
--------------------------
 
(1 row)

-- verify it can be queried from mx node
SELECT * FROM fake_table;
 a | b 
---+---
(0 rows)

SELECT * FROM run_command_on_workers('SELECT count(*) FROM fake_table')
ORDER BY nodeport;
 nodename  | nodeport | success | result 
-----------+----------+---------+--------
 localhost |    57637 | t       | 0
 localhost |    57638 | t       | 0
(2 rows)

-- create secont fake table
CREATE FOREIGN TABLE second_fake_table(a int, b int) SERVER fake_fdw_server;
SELECT create_distributed_table('second_fake_table', 'a');
NOTICE:  foreign-data wrapper "fake_fdw" does not have an extension defined
NOTICE:  foreign-data wrapper "fake_fdw" does not have an extension defined
 create_distributed_table 
--------------------------
 
(1 row)

-- verify it can be queried from mx nodes
SELECT * FROM second_fake_table;
 a | b 
---+---
(0 rows)

SELECT * FROM run_command_on_workers('SELECT count(*) FROM second_fake_table')
ORDER BY nodeport;
 nodename  | nodeport | success | result 
-----------+----------+---------+--------
 localhost |    57637 | t       | 0
 localhost |    57638 | t       | 0
(2 rows)

-- cleanup resources
DROP FUNCTION IF EXISTS fake_fdw_handler() CASCADE;
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to foreign-data wrapper fake_fdw
drop cascades to server fake_fdw_server
drop cascades to foreign table fake_table
drop cascades to foreign table second_fake_table
SELECT * FROM run_command_on_workers('DROP FUNCTION IF EXISTS fake_fdw_handler() CASCADE')
ORDER BY nodeport;
 nodename  | nodeport | success |    result     
-----------+----------+---------+---------------
 localhost |    57637 | t       | DROP FUNCTION
 localhost |    57638 | t       | DROP FUNCTION
(2 rows)

